src/log-store/src/kafka/util/record.rs

```diff
@@ -0,0 +1,669 @@
+            entries: Vec::new(),
+        }
+    }
+
+    /// Populates the entry buffer with the given entries.
+    pub(crate) fn with_entries(self, entries: Vec<EntryImpl>) -> Self {
+        Self { entries, ..self }
+    }
+
+    /// Pushes an entry into the entry buffer.
+    pub(crate) fn push(&mut self, entry: EntryImpl) {
+        self.entries.push(entry);
+    }
+
+    /// Produces the buffered entries to Kafka sever. Those entries may span several Kafka records.
+    pub(crate) async fn produce(self, client_manager: &ClientManagerRef) -> Result<Offset> {
+        ensure!(!self.entries.is_empty(), EmptyEntriesSnafu);
+
+        // Gets the producer in which a record buffer is maintained.
+        let producer = client_manager
+            .get_or_insert(&self.ns.topic)
+            .await
+            .map_err(|e| {
+                GetClientSnafu {
+                    topic: &self.ns.topic,
+                    error: e.to_string(),
+                }
+                .build()
+            })?
+            .producer;
+
+        // Stores the offset of the last successfully produced record.
+        let mut last_offset = None;
+        for entry in self.entries {
+            for record in build_records(entry, self.max_record_size >> 1) {
+                let kafka_record: KafkaRecord = record.try_into()?;
+                // Records of a certain region cannot be produced in parallel since their order must be static.
+                let offset = producer.produce(kafka_record).await.map(Offset).context(
+                    ProduceRecordSnafu {
+                        topic: &self.ns.topic,
+                    },
+                )?;
+                last_offset = Some(offset);
+            }
+        }
+        // Safety: there must be at least one record produced when the entries are guaranteed not empty.
+        Ok(last_offset.unwrap())
+    }
+}
+
+fn check_records(records: &[Record]) -> Result<()> {
+    let len = records.len();
+    ensure!(
+        len > 0,
+        IllegalSequenceSnafu {
+            error: "Empty sequence"
+        }
+    );
+
+    let mut sequence = Vec::with_capacity(len);
+    let mut entry_ids = HashSet::with_capacity(len);
+    let mut namespaces = HashSet::with_capacity(len);
+    let mut checksum_matched = HashSet::with_capacity(len);
+    for record in records {
+        sequence.push(i32::from(record.tp));
+        entry_ids.insert(record.entry_id);
+        namespaces.insert(&record.ns);
+        checksum_matched.insert(record.checksum == crc32fast::hash(&record.data));
+    }
+
+    ensure!(
+        entry_ids.len() == 1,
+        IllegalSequenceSnafu {
+            error: "Non-unique entry ids"
+        }
+    );
+
+    ensure!(
+        namespaces.len() == 1,
+        IllegalSequenceSnafu {
+            error: "Non-unique namespaces"
+        }
+    );
+
+    ensure!(
+        checksum_matched.len() == 1 && checksum_matched.contains(&true),
+        ChecksumMismatchedSnafu
+    );
```

Comment:
```suggestion
    let unique_entry_id = records.windows(2).all(|r| r[0].entry_id == r[1].entry_id);
    let unique_ns = records.windows(2).all(|r| r[0].ns == r[1].ns);
    let checksum_matched = records
        .iter()
        .all(|record| record.checksum == crc32fast::hash(&record.data));
    let sequence = records
        .into_iter()
        .map(|record| record.tp.into())
        .collect::<Vec<i32>>();

    ensure!(
        unique_entry_id,
        IllegalSequenceSnafu {
            error: "Non-unique entry ids"
        }
    );

    ensure!(
        unique_ns,
        IllegalSequenceSnafu {
            error: "Non-unique namespaces"
        }
    );

    ensure!(checksum_matched, ChecksumMismatchedSnafu);
```

Reply:
Checking records is now performed while consuming records and it's refactored.